python/arcticdb/version_store/_normalization.py

```diff
@@ -660,7 +698,10 @@ def denormalize(self, item, norm_meta):
         columns, denormed_columns, data = _denormalize_columns(item, norm_meta, idx_type, n_indexes)
 
         if not self._skip_df_consolidation:
-            df = DataFrame(data, index=index, columns=columns)
+            # Specifying the column dtype since pandas might just convert the type of empty columns to something
+            # else if we don't specify explicitly cast it depending on the version.
+            columns_dtype = None if data is None else {name: np_array.dtype for name, np_array in data.items()}
+            df = DataFrame(data, index=index, columns=columns).astype(dtype=columns_dtype)
```

Comment:
Use the `dtype` parameter of the `DataFrame` constructor. Otherwise, `astype` does a very expensive copy.

Reply:
IIRC, I tried, but the `dtype` keyword does not respect the specification and it is converting them to something else. This horrible solution is the first one I have found to respect the specification of dtypes.